src/System.Slices/System/Buffers/OwnedMemory.cs

```diff
@@ -74,118 +110,176 @@ protected void Initialize(T[] array, int arrayOffset, int length, IntPtr pointer
+        /// <summary>
+        /// Override this if you need notification of when pooling the <seealso cref="OwnedMemory{T}"/> itself is safe. 
+        /// Its references will be cleared by this point, so use <seealso cref="Dispose(bool)"/> if you want to pool its data.
+        /// </summary>
+        protected virtual void DisposeComplete()
+        { }
+
+        public bool IsDisposed => _versionId == InitializedId;
 
-        public void AddReference()
+        internal bool IsDependancyDisposed(long versionId, int reservationId)
         {
-            OnReferenceCountChanged(Interlocked.Increment(ref _references));
+            return (_versionId != versionId || IsReservationDisposed(reservationId));
         }
 
-        public void Release()
+        private bool IsReservationDisposed(int reservationId)
         {
-            OnReferenceCountChanged(Interlocked.Decrement(ref _references));
+            var index = reservationId >> 5; // divide 32
+            var offset = (uint)(1 << (reservationId & 31));
+
+            return (_outstandingReferences[index] & offset) == 0;
         }
 
-        protected virtual void OnReferenceCountChanged(int newReferenceCount)
-        { }
+        private int DisposeReservation(int reservationId)
+        {
+            var index = reservationId >> 5; // divide 32
+            var offset = (uint)(1 << (reservationId & 31));
+            var current = _outstandingReferences[index];
 
-        protected internal virtual DisposableReservation Reserve(ref ReadOnlyMemory<T> memory)
+            if ((current & offset) != 0) {
+                _outstandingReferences[index] = current & ~offset;
+                return Interlocked.Decrement(ref _references);
+            }
+            return -1;
+        }
+
+        protected internal virtual ReservedMemory<T> Reserve(ref Memory<T> memory, long versionId, int reservationId)
         {
-            return new DisposableReservation(this, Id);
+            return new ReservedMemory<T>(ref memory, this, versionId, reservationId);
         }
 
+        protected internal virtual ReservedReadOnlyMemory<T> Reserve(ref ReadOnlyMemory<T> memory, long versionId, int reservationId)
+        {
+            return new ReservedReadOnlyMemory<T>(ref memory, this, versionId, reservationId);
+        }
         #endregion
 
         #region Used by Memory<T>
-        void IKnown.AddReference(long id)
+        internal int AddReference(long versionId, int reservationId)
         {
-            VerifyId(id);
-            AddReference();
+            ThrowIfDisposed(versionId, reservationId);
+            Interlocked.Increment(ref _references);
+            var newReservationId = Interlocked.Increment(ref _lastReference);
+            var index = newReservationId >> 5; // divide 32
+            var offset = (uint)(1 << (newReservationId & 31));
+
+            if (index > _outstandingReferences.Length)
+            {
+                var refs = new uint[index + 1];
```

Comment:
_outstandingReferences is just a counted bit mask; there is an issue with `Empty` OwnedMemory as it would keep growing, but that has disposal issues anyway.

Idea in this pattern is if `_outstandingReferences` array > 1 in length on disposal it is released and set to empty. If it is length == 1 it is kept and cleared; so in reuse if less than 32 reservations were made on it there would be no allocations.